File: demo/fabric-evolution/run-fabric-evolution-latrot.py
# N. M. Rathmann <rathmann@nbi.ku.dk>, 2020-2023

import sys, os, code # code.interact(local=locals())

import numpy as np
import scipy.special as sp
from netCDF4 import Dataset

from specfabpy import specfab as sf
from specfabpy import integrator as sfint
from specfabpy import constants as sfconst

#----------------------
# Experiment selection
#----------------------

if len(sys.argv) != 2:
    print('usage: python3 %s ugrad'%(sys.argv[0]))
    sys.exit(0)
    
exp     = sys.argv[1]
exptype = exp[0:2]
ijstr   = exp[-2:]

ii_to_axis  = lambda ij: 0 if (ij=='xx') else (1 if (ij=='yy') else 2)
ij_to_plane = lambda ij: 0 if (ij=='yz') else (1 if (ij=='xz') else 2)

if exptype == 'ue': # Uniaxial extension
    DK = dict(type='ps', tau=-1, q=0,  axis=ii_to_axis(ijstr))
    strain_target = 3

if exptype == 'uc': # Uniaxial compression
    DK = dict(type='ps', tau=1, q=0,  axis=ii_to_axis(ijstr))
    strain_target = -0.98

if exptype == 'cc': # Confined compression
    DK = dict(type='ps', tau=1, q=+1, axis=ii_to_axis(ijstr))
    strain_target = -0.98

if exptype == 'ss': # Simple shear
    DK = dict(type='ss', tau=1, plane=ij_to_plane(ijstr))
    strain_target = np.deg2rad(70)
    
if exptype == 'rr': # Ridgid rotation
    DK = dict(type='rr', tau=1, plane=ij_to_plane(ijstr))
    strain_target = np.deg2rad(90)

#----------------------
# Model integration
#----------------------

Nt = 50 # Number of time steps
L  = 8 # Spectral truncation

lm, nlm_len = sf.init(L)
nlm, F, time, ugrad = sfint.lagrangianparcel(sf, DK, strain_target, Nt=Nt, iota=1, nu=1)

#----------------------
# Determine eigenvalues, principal directions, and enhancement factors
#----------------------

# Empty structure to fill
vecdim = (Nt+1,3)
eigvals  = np.zeros(vecdim)
m1,m2,m3 = np.zeros(vecdim),np.zeros(vecdim),np.zeros(vecdim)
p1,p2,p3 = np.zeros(vecdim),np.zeros(vecdim),np.zeros(vecdim)
vecdim = (Nt+1,6)
Eij_lin, Eij_nlin   = np.zeros(vecdim), np.zeros(vecdim)
Epij_lin, Epij_nlin = np.zeros(vecdim), np.zeros(vecdim)

# Grain parameters
(Eij_grain_lin,  alpha_lin,  n_grain_lin)  = sfconst.ice['viscoplastic']['linear']
(Eij_grain_nlin, alpha_nlin, n_grain_nlin) = sfconst.ice['viscoplastic']['nonlinear']

viscparams = (Eij_grain_lin, alpha_lin, n_grain_lin)

for tt in np.arange(0,Nt+1):

    c = nlm[tt,:]
    
    m1[tt,:],m2[tt,:],m3[tt,:], eigvals[tt,:] = sf.frame(c, 'e')
    p1[tt,:],p2[tt,:],p3[tt,:], _             = sf.frame(c, 'p')

    # Linear (n'=1) mixed Taylor--Sachs enhancements            
    Eij_lin[tt,:]  = sf.Eij_tranisotropic(c, m1[tt,:],m2[tt,:],m3[tt,:], Eij_grain_lin, alpha_lin, n_grain_lin)
    Epij_lin[tt,:] = sf.Eij_tranisotropic(c, p1[tt,:],p2[tt,:],p3[tt,:], Eij_grain_lin, alpha_lin, n_grain_lin)
    
    # Nonlinear (n'=3) Sachs enhancements
    Eij_nlin[tt,:]  = sf.Eij_tranisotropic(c, m1[tt,:],m2[tt,:],m3[tt,:], Eij_grain_nlin, alpha_nlin, n_grain_nlin)
    Epij_nlin[tt,:] = sf.Eij_tranisotropic(c, p1[tt,:],p2[tt,:],p3[tt,:], Eij_grain_nlin, alpha_nlin, n_grain_nlin)
    
#----------------------
# Save
#----------------------

fname = 'solutions/LROT_%s.nc'%(exp)
ncfile = Dataset(fname,mode='w',format='NETCDF3_CLASSIC') 

# Config
dt = time[1]-time[0]
ncfile.tsteps, ncfile.dt, ncfile.L = Nt, dt, L
ncfile.Ecc_lin,  ncfile.Eca_lin,  ncfile.alpha_lin  = Eij_grain_lin[0],  Eij_grain_lin[1],  alpha_lin
ncfile.Ecc_nlin, ncfile.Eca_nlin, ncfile.alpha_nlin = Eij_grain_nlin[0], Eij_grain_nlin[1], alpha_nlin

ncfile.ugrad = ugrad.flatten()

# Dimensions
c_did       = ncfile.createDimension('DOF',     nlm_len)
time_did    = ncfile.createDimension('tstep',   Nt+1)
eig_did     = ncfile.createDimension('eigval',  3)
dim_did     = ncfile.createDimension('dim',     3)
dim6_did    = ncfile.createDimension('dim6',    6)
pair_did    = ncfile.createDimension('pair',    2)

# Variables
myint = np.int32
myflt = np.float64
dimarr_vec = ('lm','lm','lmdyn')

f_lm   = ncfile.createVariable('lm', myint, ('DOF','pair')) 
f_c_re = ncfile.createVariable('c_re', myflt, ('tstep','DOF')) 
f_c_im = ncfile.createVariable('c_im', myflt, ('tstep','DOF')) 
f_eigvals = ncfile.createVariable('eigvals', myflt, ('tstep','eigval'))

f_Eij_lin   = ncfile.createVariable('Eij_lin',   myflt, ('tstep','dim6'))
f_Epij_lin  = ncfile.createVariable('Epij_lin',  myflt, ('tstep','dim6'))
f_Eij_nlin  = ncfile.createVariable('Eij_nlin',  myflt, ('tstep','dim6'))
f_Epij_nlin = ncfile.createVariable('Epij_nlin', myflt, ('tstep','dim6'))

mkvec = lambda field: ncfile.createVariable(field, myflt, ('tstep','dim'))
f_m1, f_m2, f_m3 = mkvec('m1'),  mkvec('m2'),  mkvec('m3')
f_p1, f_p2, f_p3 = mkvec('p1'),  mkvec('p2'),  mkvec('p3')

f_lm[:,:], f_c_re[:,:], f_c_im[:,:] = lm.T, np.real(nlm), np.imag(nlm)
f_Eij_lin[:,:],  f_Eij_nlin[:,:]  = Eij_lin,  Eij_nlin
f_Epij_lin[:,:], f_Epij_nlin[:,:] = Epij_lin, Epij_nlin
f_m1[:,:], f_m2[:,:], f_m3[:,:] = m1, m2, m3
f_p1[:,:], f_p2[:,:], f_p3[:,:] = p1, p2, p3
f_eigvals[:,:] = eigvals

ncfile.close(); 
print('Solution dumped in %s'%fname)
print('Plot result: python3 plot-fabric-evolution-latrot.py %s'%(exp))

File: demo/fabric-evolution/run-fabric-evolution-latrot.f90
! N. M. Rathmann <rathmann@nbi.ku.dk> and D. A. Lilien <dlilien90@gmail.com>, 2020

program demo

    use specfab    
    use netcdf
    
    implicit none

    integer, parameter :: dp = 8

    ! Numerics
    integer, parameter :: Nt = 50  ! Number of time steps
    real, parameter    :: dt = 0.0782404601085629 ! Time-step size (gives a vertical strain of -0.98 for experiment "uc_zz")
    integer            :: Lcap = 8 ! Spectral truncation (4<=L<=8)
    
    ! Constants and argv strings    
    integer          :: ii,tt ! loop vars
    character(len=5) :: arg_exp ! experiment type (see below)

    ! Fabric state and evolution
    complex(kind=dp), allocatable :: nlm(:), dndt(:,:), dndt_ROT(:,:), dndt_REG(:,:) ! Series expansion coefs and evolution matrix
    real(kind=dp)                 :: ugrad(3,3), eps(3,3), omg(3,3) ! Large-scale deformation

    ! For dumping state to netCDF
    complex(kind=dp), allocatable   :: nlm_save(:,:)
    real(kind=dp)                   :: eigvals_save(3,Nt), Eeiej_lin_save(3,3,Nt), Eeiej_nlin_save(3,3,Nt), Epipj_lin_save(3,3,Nt), Epipj_nlin_save(3,3,Nt)
    real(kind=dp), dimension(3,Nt)  :: e1_save,e2_save,e3_save, p1_save,p2_save,p3_save
    character(len=30) :: fname_sol
    integer :: ncid, c_did, time_did, eig_did, dim_did, pair_did ! Dimension IDs
    integer :: id_cre,id_cim,id_lm, id_eig, id_e1,id_e2,id_e3, id_p1,id_p2,id_p3! Var IDs
    integer :: id_Eeiej_lin, id_Eeiej_nlin, id_Epipj_lin, id_Epipj_nlin ! Var IDs

    if (command_argument_count() .ne. 1) then
        print *,'usage: ./demo ugrad'
        call exit(0)
    end if

    !-------------------------------------------------------------------
    ! Velocity gradient tensor
    !-------------------------------------------------------------------

    call get_command_argument(1, arg_exp)
    select case (arg_exp)

        ! RECALL COLUMN FIRST IN FORTRAN
        
        ! Uniaxial compression (uc) and uniaxial extension (ue)
        case ('uc_xx', 'ue_xx')
            ugrad = reshape([-1.0,0.,0., 0.,0.5,0., 0.,0.,0.5], [3, 3])
        case ('uc_yy', 'ue_yy')
            ugrad = reshape([0.5,0.,0., 0.,-1.0,0., 0.,0.,0.5], [3, 3])
        case ('uc_zz', 'ue_zz')
            ugrad = reshape([0.5,0.,0., 0.,0.5,0., 0.,0.,-1.], [3, 3])

        ! Confined compression (cc)
        case ('cc_zx')
            ugrad = reshape([1.,0.,0., 0.,0.,0., 0.,0.,-1.], [3, 3]) ! confined in y
        case ('cc_zy')
            ugrad = reshape([0.,0.,0., 0.,1.,0., 0.,0.,-1.], [3, 3]) ! confined in x
        case ('cc_yx')
            ugrad = reshape([1.,0.,0., 0.,-1.,0., 0.,0.,0.], [3, 3]) ! confined in z
                    
        ! Simple shear (ss)
        case ('ss_xz')
            ugrad = reshape([0.,0.,0., 0.,0.,0., 1.,0.,0.], [3, 3]) 
        case ('ss_xy')
            ugrad = reshape([0.,0.,0., 1.,0.,0., 0.,0.,0.], [3, 3]) 
        case ('ss_yz')
            ugrad = reshape([0.,0.,0., 0.,0.,0., 0.,1.,0.], [3, 3]) 

        ! Ridig rotation (rr)
        case ('rr_xz')
            ugrad = reshape([0.,0.,-1., 0.,0.,0., +1.,0.,0.], [3, 3])
        case ('rr_xy')
            ugrad = reshape([0.,-1.,0., +1.,0.,0., 0.,0.,0.], [3, 3]) 
        case ('rr_yz')
            ugrad = reshape([0.,0.,0., 0.,0.,-1., 0.,+1.,0.], [3, 3]) 

        case default
            print *,'argv error: valid "ugrad" are "uc_ij", "ue_ij", "cc_ij", "ss_ij", "rr_ij" where i,j are any of x,y,z'
            call exit(0)
    end select

    select case (arg_exp)
        case ('ue_xx','ue_yy','ue_zz')
            ugrad = -1*ugrad
    end select

    eps = (ugrad+transpose(ugrad))/2 ! strain-rate
    omg = (ugrad-transpose(ugrad))/2 ! spin
            
    !-------------------------------------------------------------------
    ! Initialize
    !-------------------------------------------------------------------

    call initspecfab(Lcap) ! nlm_len is now defined (number of expansion coeffcients, i.e. #DOFs)
    allocate(nlm(nlm_len))
    nlm = [(0,ii=1,nlm_len)] ! Expansion coefs "n_l^m" are saved in the 1D array "nlm". Corresponding (l,m) values for the i'th coef (i.e. nlm(i)) are (l,m) = (lm(1,i),lm(2,i))
    allocate(nlm_save(nlm_len,Nt))
    allocate(dndt(nlm_len,nlm_len))
    allocate(dndt_ROT(nlm_len,nlm_len))
    allocate(dndt_REG(nlm_len,nlm_len))    

    
    select case (arg_exp)
        case ('rr_xz','rr_xy','rr_yz')
            ! Init with some anisotropy
            nlm(1) = sqrt(cmplx(1/2.,0))
            nlm(3) = sqrt(cmplx(1/2.,0))
        case default
            ! Init with isotropy
            nlm(1) = (1,0)
            nlm(1) = nlm(1)/f_ev_c0(nlm(1)) ! Normalize
    end select
    
    write(*,"(A13,I4,A5,F12.10,A4,I2,A10,I3,A1)") 'Numerics: Nt=', Nt, ', dt=', dt, ', L=', Lcap, ' (nlm_len=',nlm_len,')'

    !-------------------------------------------------------------------
    ! Integrate
    !-------------------------------------------------------------------

    call savestate(nlm, 1) ! Save initial state    
    dndt_ROT = dndt_ij_LATROT(eps,omg, 0*eps,0d0,0d0,0d0, 1d0) ! Assume constant strain-rate and spin with Taylor style plastic spin for lattice rotation (beta=1).
    dndt_REG = dndt_ij_REG(eps) ! Regularization
!    dndt_REG = 5 * dndt_ij_CDRX() ! Rotation recrystalization
            
    do tt = 2, Nt
!        write(*,"(A9,I3)") '*** Step ', tt
        dndt = dndt_ROT + dndt_REG 
        nlm = nlm + dt * matmul(dndt, nlm) ! Spectral coefficients evolve by a linear transformation
        call savestate(nlm, tt)
    end do
    
    !-------------------------------------------------------------------
    ! Dump solution to netCDF
    !-------------------------------------------------------------------
    
    write (fname_sol,"('solutions/LATROT_',A5,'.nc')") arg_exp
    call check( nf90_create(fname_sol, NF90_CLOBBER, ncid) )
    
    call check(nf90_put_att(ncid,NF90_GLOBAL, "tsteps", Nt))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "dt",     dt))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "L",      Lcap))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "ugrad",  reshape(ugrad, [size(ugrad)]) ))
    
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Eca_lin", Eca_opt_lin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Ecc_lin", Ecc_opt_lin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Eca_nlin", Eca_opt_nlin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "Ecc_nlin", Ecc_opt_nlin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "alpha_lin",  alpha_opt_lin))
    call check(nf90_put_att(ncid,NF90_GLOBAL, "alpha_nlin", alpha_opt_nlin))
    
    call check( nf90_def_dim(ncid, "DOF",    nlm_len,   c_did) )
    call check( nf90_def_dim(ncid, "tstep",  Nt,        time_did) )
    call check( nf90_def_dim(ncid, "eigval", 3,         eig_did) )
    call check( nf90_def_dim(ncid, "dim",    3,         dim_did) )
    call check( nf90_def_dim(ncid, "pair",   2,         pair_did) )

    call check( nf90_def_var(ncid, "lm",    NF90_INT,    [pair_did, c_did], id_lm) )
    call check( nf90_def_var(ncid, "c_re",  NF90_DOUBLE, [c_did,   time_did], id_cre) )
    call check( nf90_def_var(ncid, "c_im",  NF90_DOUBLE, [c_did,   time_did], id_cim) )
    
    call check( nf90_def_var(ncid, "eigvals", NF90_DOUBLE, [eig_did, time_did], id_eig) )
    call check( nf90_def_var(ncid, "Eeiej_nlin", NF90_DOUBLE, [dim_did,dim_did, time_did], id_Eeiej_nlin) )
    call check( nf90_def_var(ncid, "Eeiej_lin",  NF90_DOUBLE, [dim_did,dim_did, time_did], id_Eeiej_lin) )
    call check( nf90_def_var(ncid, "Epipj_nlin", NF90_DOUBLE, [dim_did,dim_did, time_did], id_Epipj_nlin) )
    call check( nf90_def_var(ncid, "Epipj_lin",  NF90_DOUBLE, [dim_did,dim_did, time_did], id_Epipj_lin) )
    call check( nf90_def_var(ncid, "e1", NF90_DOUBLE, [dim_did, time_did], id_e1) )
    call check( nf90_def_var(ncid, "e2", NF90_DOUBLE, [dim_did, time_did], id_e2) )
    call check( nf90_def_var(ncid, "e3", NF90_DOUBLE, [dim_did, time_did], id_e3) )
    call check( nf90_def_var(ncid, "p1", NF90_DOUBLE, [dim_did, time_did], id_p1) )
    call check( nf90_def_var(ncid, "p2", NF90_DOUBLE, [dim_did, time_did], id_p2) )
    call check( nf90_def_var(ncid, "p3", NF90_DOUBLE, [dim_did, time_did], id_p3) )

    call check( nf90_enddef(ncid) )
    
    call check( nf90_put_var(ncid, id_cre,   real(nlm_save)) )
    call check( nf90_put_var(ncid, id_cim,   aimag(nlm_save)) )
    call check( nf90_put_var(ncid, id_lm,    lm(:,1:nlm_len)) )
    
    call check( nf90_put_var(ncid, id_eig,  eigvals_save) )
    call check( nf90_put_var(ncid, id_Eeiej_lin,  Eeiej_lin_save) )
    call check( nf90_put_var(ncid, id_Eeiej_nlin, Eeiej_nlin_save) )
    call check( nf90_put_var(ncid, id_e1, e1_save) )
    call check( nf90_put_var(ncid, id_e2, e2_save) )
    call check( nf90_put_var(ncid, id_e3, e3_save) )
    call check( nf90_put_var(ncid, id_Epipj_lin,  Epipj_lin_save) )
    call check( nf90_put_var(ncid, id_Epipj_nlin, Epipj_nlin_save) )
    call check( nf90_put_var(ncid, id_p1, p1_save) )
    call check( nf90_put_var(ncid, id_p2, p2_save) )
    call check( nf90_put_var(ncid, id_p3, p3_save) )
    
    call check( nf90_close(ncid) )

    print *, 'Solution dumped in ', fname_sol
    print *, "Plot result:"
    write(*,"(A41,A5)") "python3 plot-fabric-evolution-latrot.py ", arg_exp

contains

    subroutine savestate(nlm, tt)
    
        implicit none
    
        complex(kind=dp), intent(in) :: nlm(nlm_len)
        integer, intent(in) :: tt
    
        nlm_save(:,tt)  = nlm

        call frame(nlm, 'e', e1_save(:,tt),e2_save(:,tt),e3_save(:,tt), eigvals_save(:,tt)) ! eigen frame 
        call frame(nlm, 'p', p1_save(:,tt),p2_save(:,tt),p3_save(:,tt), eigvals_save(:,tt)) ! eigen frame rotated 45 deg. (pq-frame)

        ! Linear (n'=1) mixed Taylor--Sachs enhancements        
        Eeiej_lin_save(:,:,tt) = Eeiej(nlm, e1_save(:,tt),e2_save(:,tt),e3_save(:,tt), Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, 1)
        Epipj_lin_save(:,:,tt) = Eeiej(nlm, p1_save(:,tt),p2_save(:,tt),p3_save(:,tt), Ecc_opt_lin, Eca_opt_lin, alpha_opt_lin, 1)
        
        ! Nonlinear (n'=3) Sachs enhancements
        Eeiej_nlin_save(:,:,tt) = Eeiej(nlm, e1_save(:,tt),e2_save(:,tt),e3_save(:,tt), Ecc_opt_nlin, Eca_opt_nlin, alpha_opt_nlin, 3)
        Epipj_nlin_save(:,:,tt) = Eeiej(nlm, p1_save(:,tt),p2_save(:,tt),p3_save(:,tt), Ecc_opt_nlin, Eca_opt_nlin, alpha_opt_nlin, 3)

    end
    
    subroutine check(status)
        implicit none
        integer, intent (in) :: status
        if(status /= nf90_noerr) then 
            print *, trim(nf90_strerror(status))
            stop "Stopped"
        end if
    end subroutine check  
    
end program

